File: rust/src/rate.rs
use std::f64::consts::PI;

use crate::constants::HBAR_EV;
use crate::cotunneling::{sum_mmmmrs, sum_mmmmrs11, sum_mmr, sum_mmr11};
use crate::fc_matrix::FCCache;
use crate::fermi_bose::fermi;

/// Flat-array rate store indexed by [n1][n2][q1][lead_idx][q2].
pub struct RateStore {
    pub n: usize,
    pub data: Vec<f64>,
}

impl RateStore {
    pub fn new(n: usize) -> Self {
        RateStore {
            n,
            data: vec![0.0; 4 * n * 2 * n],
        }
    }

    #[inline]
    fn lead_to_idx(lead: i32) -> usize {
        if lead == 1 { 0 } else { 1 }
    }

    #[inline]
    fn rate_idx(n: usize, n1: usize, n2: usize, q1: usize, lead_idx: usize, q2: usize) -> usize {
        (((n1 * 2 + n2) * n + q1) * 2 + lead_idx) * n + q2
    }

    #[inline]
    pub fn rate_w(&self, n1: usize, n2: usize, q1: usize, q2: usize, lead: i32) -> f64 {
        if n1 > 1 || n2 > 1 || q1 >= self.n || q2 >= self.n {
            return 0.0;
        }
        let lead_idx = Self::lead_to_idx(lead);
        let idx = Self::rate_idx(self.n, n1, n2, q1, lead_idx, q2);
        self.data[idx]
    }

    #[inline]
    pub fn rate_w_lead(&self, n1: usize, n2: usize, q1: usize, q2: usize) -> f64 {
        self.rate_w(n1, n2, q1, q2, 1) + self.rate_w(n1, n2, q1, q2, -1)
    }
}

#[inline(always)]
fn sanitize(x: f64) -> f64 {
    if x.is_finite() { x } else { 0.0 }
}

pub fn spin_degeneracy(n1: usize, n2: usize) -> f64 {
    if n1 == 0 {
        2.0
    } else if n2 == 0 {
        1.0
    } else {
        2.0
    }
}

/// Compute rates for all q2 given (n1, n2, q1, lead).
pub fn m_rate_w(
    n1: usize,
    n2: usize,
    q1: i32,
    q2_vec: &[i32],
    vmode: f64,
    alpha_l: f64,
    alpha_r: f64,
    lambda: f64,
    vsd: f64,
    t: f64,
    eta: f64,
    lead: i32,
    vg: f64,
    fc: &mut FCCache,
    out: &mut [f64],
) {
    let nq2 = q2_vec.len();
    let gamma_l = alpha_l * vmode;
    let gamma_r = alpha_r * vmode;
    let epsilond = 0.0 + vg;
    let mu_l = eta * vsd;
    let mu_r = -(1.0 - eta) * vsd;

    let s = spin_degeneracy(n1, n2);
    let gamma = if lead == 1 { gamma_l } else { gamma_r };
    let mu = if lead == 1 { mu_l } else { mu_r };

    if lead != 1 && lead != -1 {
        for v in out.iter_mut().take(nq2) {
            *v = 0.0;
        }
        return;
    }

    if n1 == 1 && n2 == 0 {
        // Sequential 1→0
        for i in 0..nq2 {
            let q2 = q2_vec[i];
            let fc_val = fc.get(q1, q2);
            let f = fermi(epsilond - (q2 - q1) as f64 * vmode, mu, t);
            let val = s * gamma / HBAR_EV * fc_val * fc_val * (1.0 - f);
            out[i] = sanitize(val);
        }
    } else if n1 == 0 && n2 == 1 {
        // Sequential 0→1
        for i in 0..nq2 {
            let q2 = q2_vec[i];
            let fc_val = fc.get(q1, q2);
            let f = fermi(epsilond + (q2 - q1) as f64 * vmode, mu, t);
            let val = s * gamma / HBAR_EV * fc_val * fc_val * f;
            out[i] = sanitize(val);
        }
    } else if n1 == 0 && n2 == 0 {
        // Cotunneling 0→0
        let mut single_sum = vec![0.0; nq2];
        let mut double_sum = vec![0.0; nq2];

        if lead == 1 {
            sum_mmr(q1, q2_vec, lambda, mu_l, mu_r, vmode, epsilond, t, fc, &mut single_sum);
            sum_mmmmrs(q1, q2_vec, lambda, mu_l, mu_r, vmode, epsilond, t, fc, &mut double_sum);
        } else {
            sum_mmr(q1, q2_vec, lambda, mu_r, mu_l, vmode, epsilond, t, fc, &mut single_sum);
            sum_mmmmrs(q1, q2_vec, lambda, mu_r, mu_l, vmode, epsilond, t, fc, &mut double_sum);
        }

        let prefac = s / (2.0 * PI * HBAR_EV) * gamma_l * gamma_r;
        for i in 0..nq2 {
            out[i] = sanitize(prefac * (single_sum[i] + double_sum[i]));
        }
    } else {
        // Cotunneling 1→1
        let mut single_sum = vec![0.0; nq2];
        let mut double_sum = vec![0.0; nq2];

        if lead == 1 {
            sum_mmr11(q1, q2_vec, lambda, mu_l, mu_r, vmode, epsilond, t, fc, &mut single_sum);
            sum_mmmmrs11(
                q1, q2_vec, lambda, mu_l, mu_r, vmode, epsilond, t, fc, &mut double_sum,
            );
        } else {
            sum_mmr11(q1, q2_vec, lambda, mu_r, mu_l, vmode, epsilond, t, fc, &mut single_sum);
            sum_mmmmrs11(
                q1, q2_vec, lambda, mu_r, mu_l, vmode, epsilond, t, fc, &mut double_sum,
            );
        }

        let prefac = s / (2.0 * PI * HBAR_EV) * gamma_l * gamma_r;
        for i in 0..nq2 {
            out[i] = sanitize(prefac * (single_sum[i] + double_sum[i]));
        }
    }
}

/// Pre-compute all rates for a given lead and store them.
pub fn calculate_all_rate_w(
    store: &mut RateStore,
    n: usize,
    vmode: f64,
    alpha_l: f64,
    alpha_r: f64,
    lambda: f64,
    vsd: f64,
    t: f64,
    eta: f64,
    lead: i32,
    vg: f64,
    fc: &mut FCCache,
) {
    let q2_vec: Vec<i32> = (0..n as i32).collect();
    let mut out_buf = vec![0.0; n];
    let lead_idx = RateStore::lead_to_idx(lead);

    for n1 in 0..=1_usize {
        for n2 in 0..=1_usize {
            for q1 in 0..n {
                m_rate_w(
                    n1, n2, q1 as i32, &q2_vec, vmode, alpha_l, alpha_r, lambda, vsd, t, eta,
                    lead, vg, fc, &mut out_buf,
                );
                for q2 in 0..n {
                    let idx = RateStore::rate_idx(n, n1, n2, q1, lead_idx, q2);
                    store.data[idx] = out_buf[q2];
                }
            }
        }
    }
}

File: cpp/src/rate.cpp
#include "rate.hpp"

#include <cmath>
#include <vector>

#include "constants.hpp"
#include "cotunneling.hpp"
#include "fermi_bose.hpp"

namespace fc {

static constexpr double PI = 3.14159265358979323846;

static inline int lead_to_idx(int lead) {
    return (lead == 1) ? 0 : 1;
}

static inline int rate_idx(int N, int n1, int n2, int q1, int lead_idx, int q2) {
    return ((((n1 * 2 + n2) * N + q1) * 2 + lead_idx) * N + q2);
}

static inline double sanitize(double x) {
    return std::isfinite(x) ? x : 0.0;
}

RateStore::RateStore(int N) : N(N), data(4 * N * 2 * N, 0.0) {}

double spin_degeneracy(int n1, int n2) {
    if (n1 == 0) {
        return 2.0;
    }
    return (n2 == 0) ? 1.0 : 2.0;
}

void m_rateW(int n1, int n2, int q1, const int* q2_vec, int nq2,
             double vmode, double alphaL, double alphaR, double lambda,
             double Vsd, double T, double eta, int lead, double Vg,
             const FCCache& fc, double* out) {
    (void)lambda;
    if (nq2 <= 0) {
        return;
    }

    const double gammaL = alphaL * vmode;
    const double gammaR = alphaR * vmode;
    const double epsilond = 0.0 + Vg;
    const double muL = eta * Vsd;
    const double muR = -(1.0 - eta) * Vsd;

    const double s = spin_degeneracy(n1, n2);
    const double gamma = (lead == 1) ? gammaL : gammaR;
    const double mu = (lead == 1) ? muL : muR;

    if (lead != 1 && lead != -1) {
        for (int i = 0; i < nq2; ++i) {
            out[i] = 0.0;
        }
        return;
    }

    if (n1 == 1 && n2 == 0) {
        for (int i = 0; i < nq2; ++i) {
            const int q2 = q2_vec[i];
            const double fc_val = fc_cache_get(fc, q1, q2);
            const double f = fermi(epsilond - static_cast<double>(q2 - q1) * vmode, mu, T);
            const double val = s * gamma / HBAR_EV * fc_val * fc_val * (1.0 - f);
            out[i] = sanitize(val);
        }
    } else if (n1 == 0 && n2 == 1) {
        for (int i = 0; i < nq2; ++i) {
            const int q2 = q2_vec[i];
            const double fc_val = fc_cache_get(fc, q1, q2);
            const double f = fermi(epsilond + static_cast<double>(q2 - q1) * vmode, mu, T);
            const double val = s * gamma / HBAR_EV * fc_val * fc_val * f;
            out[i] = sanitize(val);
        }
    } else if (n1 == 0 && n2 == 0) {
        std::vector<double> single_sum(nq2);
        std::vector<double> double_sum(nq2);

        if (lead == 1) {
            sumMMr(q1, q2_vec, nq2, lambda, muL, muR, vmode, epsilond, T, fc, single_sum.data());
            sumMMMMrs(q1, q2_vec, nq2, lambda, muL, muR, vmode, epsilond, T, fc, double_sum.data());
        } else {
            sumMMr(q1, q2_vec, nq2, lambda, muR, muL, vmode, epsilond, T, fc, single_sum.data());
            sumMMMMrs(q1, q2_vec, nq2, lambda, muR, muL, vmode, epsilond, T, fc, double_sum.data());
        }

        const double prefac = s / (2.0 * PI * HBAR_EV) * gammaL * gammaR;
        for (int i = 0; i < nq2; ++i) {
            out[i] = sanitize(prefac * (single_sum[i] + double_sum[i]));
        }
    } else {
        std::vector<double> single_sum(nq2);
        std::vector<double> double_sum(nq2);

        if (lead == 1) {
            sumMMr11(q1, q2_vec, nq2, lambda, muL, muR, vmode, epsilond, T, fc, single_sum.data());
            sumMMMMrs11(q1, q2_vec, nq2, lambda, muL, muR, vmode, epsilond, T, fc, double_sum.data());
        } else {
            sumMMr11(q1, q2_vec, nq2, lambda, muR, muL, vmode, epsilond, T, fc, single_sum.data());
            sumMMMMrs11(q1, q2_vec, nq2, lambda, muR, muL, vmode, epsilond, T, fc, double_sum.data());
        }

        const double prefac = s / (2.0 * PI * HBAR_EV) * gammaL * gammaR;
        for (int i = 0; i < nq2; ++i) {
            out[i] = sanitize(prefac * (single_sum[i] + double_sum[i]));
        }
    }
}

void calculate_all_rateW(RateStore& store, int N,
                         double vmode, double alphaL, double alphaR,
                         double lambda, double Vsd, double T,
                         double eta, int lead, double Vg, const FCCache& fc) {
    if (N <= 0) {
        return;
    }

    std::vector<int> q2_vec(N);
    std::vector<double> out(N);

    for (int q2 = 0; q2 < N; ++q2) {
        q2_vec[q2] = q2;
    }

    const int lead_idx_val = lead_to_idx(lead);
    for (int n1 = 0; n1 <= 1; ++n1) {
        for (int n2 = 0; n2 <= 1; ++n2) {
            for (int q1 = 0; q1 < N; ++q1) {
                m_rateW(n1, n2, q1, q2_vec.data(), N, vmode, alphaL, alphaR, lambda,
                        Vsd, T, eta, lead, Vg, fc, out.data());
                for (int q2 = 0; q2 < N; ++q2) {
                    const int idx = rate_idx(N, n1, n2, q1, lead_idx_val, q2);
                    store.data[idx] = out[q2];
                }
            }
        }
    }
}

double rateW_from_store(const RateStore& store, int n1, int n2, int q1, int q2, int lead) {
    const int N = store.N;
    if (n1 < 0 || n1 > 1 || n2 < 0 || n2 > 1 || q1 < 0 || q1 >= N || q2 < 0 || q2 >= N) {
        return 0.0;
    }
    const int lid = lead_to_idx(lead);
    const int idx = rate_idx(N, n1, n2, q1, lid, q2);
    return store.data[idx];
}

double rateW_lead_from_store(const RateStore& store, int n1, int n2, int q1, int q2) {
    return rateW_from_store(store, n1, n2, q1, q2, 1) +
           rateW_from_store(store, n1, n2, q1, q2, -1);
}

}
